core/src/main/java/com/linecorp/armeria/server/throttling/ThrottlingStrategy.java

```diff
@@ -126,4 +152,11 @@ protected ThrottlingStrategy(@Nullable String name) {
     public String name() {
         return name;
     }
+
+    /**
+     * Returns the failure status to be used for the throttled requests.
+     */
+    protected HttpStatus failureStatus() {
```

Comment:
Instead of adding just this knob to strategy, how about adding `ThrottlingFailureHandler` similar to our auth service?

https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/auth/AuthService.java#L75

By letting it return an arbitrary response, we get more flexibility (including implementing the x-ratelimit headers if we add a token throttler).

Reply:
Ok, I can try that if all reviewers in agreement on that.
I have one problem with this proposed design though - its inability to pass the context directly from an `authorizer` to a `handler`. I experienced that in several places, including `AuthService` and `ThrottlingService`. In either case I ended up using `ServiceRequestContext`. While in case of `ThrottlingService` it was ok (but not very pretty) since I could set various `x-ratelimit` headers using `ctx.addAdditionalResponseHeader()` at `ThrottlingStrategy.accept()`, which sort of defeated the purpose of the error handler. In case of OAuth2 implementation, it was even trickier - token validation can fail in a number of ways and the RFC instructs to properly inform the client by composing specific JSon error response. Therefore setting up the headers to the context inside the `authorize()` was not very useful. What I needed then was an ability to pass some custom set of artefacts from an `authorizer` to a `handler`, including error response code, oauth2 error code and some additional elements none of which were headers.
Ideally, the interfaces like `Authorizer.authorize()` and `ThrottlingStrategy.accept()` should not be binary and should allow responding with some custom context.

Reply:
Ah that's a good point. I think passing the context into the handler makes sense and we probably want to do so for auth service too. For here, I was expecting the interface for the handler to match the signature of `onFailure` exactly.

Reply:
Yes, I understand the idea of the handler. Could you recommend something on the context passing solution?

Reply:
I may be missing the concern so let me know - I was thinking if the signature of the handler takes a `ServiceRequestContext` and returns a `HttpResponse`, which is not what the with service currently does, the context issue is solved since, using auth as an example, both the authorizer and handler have the context. So if here if both strategy has context (already does) and handler has it, it works. Let me know if this misses something.

Reply:
Assuming that there are two parties - the authorizer and the handler (consider auth example), where one party (the authorizer) makes decisions based on complex logic (token validation) and the other party (the handler) fulfills the decisions made.
Let's say the authorizer validates the token, and as a result, needs to pass following elements to the handler: HTTP error response code, OAuth error code, some reference data on the token (e.g. token expiry date). The latter two will have to be put into JSON response body. How such context gets passed with the proposed design?

Reply:
Assuming that there are two parties - the authorizer and the handler (consider auth example), where one party (the authorizer) makes decisions based on complex logic (token validation) and the other party (the handler) fulfills the decisions made.
Let's say the authorizer validates the token, and as a result, needs to pass following elements to the handler: HTTP error response code, OAuth error code, some reference data on the token (e.g. token expiry date). The latter two will have to be put into JSON response body. How such context gets passed with the proposed design?

Reply:
Via `RequestContext.attr()`? Not sure if it's nice, though.

Reply:
Made a second commit - Implemented `ThrottlingReject`/`AcceptHandler` approach, similar to `AuthService`, as suggested by @anuraaga, to fully customize throttling response rejection/acceptance.
Introduced builders for `ThrottlingService` and `ThrottlingRpcService`.
Changed default to `TOO_MANY_REQUESTS` only for `HTTP`. `RPC` still uses `SERVICE_UNAVAILABLE`.